nexus/src/db/datastore/volume.rs

```diff
@@ -166,15 +166,164 @@ impl DataStore {
         Ok(())
     }
 
+    /// Get a copy of the Volume from the database.
+    /// This action (getting a copy) will increase the generation number
+    /// of Volumes of the VolumeConstructionRequest::Volume type that have
+    /// sub_volumes of the VolumeConstructionRequest::Region type.
+    /// This generation number increase is required for Crucible to support
+    /// crash consistency.
     pub async fn volume_get(&self, volume_id: Uuid) -> LookupResult<Volume> {
```

Comment:
I find it unintuitive that a method named "get" is performing a mutation of the underlying record via a transaction.

[Generation Numbers are documented in RFD 192](https://rfd.shared.oxide.computer/rfd/192#_transactions_ctes_sagas_generation_numbers), and the usage there recommends the following:

> Generation numbers are often used for optimistic concurrency control. In a simple case, a client might SELECT a record that has a generation number, then later UPDATE that record conditional on the generation number not having changed. If the generation number has changed, the client might abort or retry the whole operation again, starting with the SELECT.

> You don’t usually need generation numbers when the SELECT and UPDATE parts can be placed in the same transaction because you can skip the SELECT altogether and issue the UPDATE, including whatever parts of the WHERE clause would have been in the SELECT.

What is happening in this PR doesn't seem to match that pattern - it appears that we're removing the ability to have a raw `SELECT`, and instead forcing all "reads" of volumes to bump the generation number.

Why are we bumping the generation number unconditionally on a `get`? Shouldn't we be only bumping this number when we actually have an update to perform?

Reply:
Yes, I agree the `volume_get` name here does not cover what happens.

The generation number is Crucible specific, and not related to RFD 192.
The [section on reconciliation in RFD 177](https://rfd.shared.oxide.computer/rfd/0177#_reconciliation) has more details that I should have added that link to help describe what we need to happen here.

The volume.data field of a Volume is actually just a String that contains a `VolumeConstructionRequest`, so we can't use it like a regular database entry.

In Crucible, the generation number serves two purposes.
1. To allow only a single Upstairs to access the three downstairs that make up a writable region.
2. To help an upstairs repair and reconcile the three downstairs after a crash (or really any upstairs starting).

For the above to work, a volumes generation number must always be increasing, and there must only be one "upstairs" with a given generation number.  In order to enforce these generation number requirements, we make Nexus (and specifically the `volume_get`) as the source and updater of a volumes generation number.

I like to think of `volume_get()` as someone requesting the latest copy of the volume data that will allow an upstairs to activate and use the crucible regions described in that data.  If another requestor comes along afterward, they will get the new latest copy of the volume data and will be able to "take over" that volume from any previous requestor.

Reply:
Chatted out-of-band on this: This operation probably should be renamed to `volume_checkout` or something similar, to imply that the call itself acts as a tiebreaker for Crucible upstairs.